src/Microsoft.Extensions.WebSockets/WebSocketConnection.cs

```diff
@@ -0,0 +1,391 @@
+﻿using System;
+using System.Binary;
+using System.Diagnostics;
+using System.Security.Cryptography;
+using System.Threading;
+using System.Threading.Tasks;
+using Channels;
+using Microsoft.Extensions.WebSockets.Internal;
+
+namespace Microsoft.Extensions.WebSockets
+{
+    /// <summary>
+    /// Provides the default implementation of <see cref="IWebSocketConnection"/>.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// This type is thread-safe under the following condition: No two threads attempt to call either
+    /// <see cref="ReceiveAsync"/> or <see cref="SendAsync"/> simultaneously. Different threads may call each method, but the same method
+    /// cannot be re-entered while it is being run in a different thread.
+    /// </para>
+    /// <para>
+    /// The general pattern of having a single thread running <see cref="ReceiveAsync"/> and a separate thread running <see cref="SendAsync"/> will
+    /// be thread-safe, as each method interacts with completely separate state.
+    /// </para>
+    /// </remarks>
+    public class WebSocketConnection : IWebSocketConnection
+    {
+        private static readonly RandomNumberGenerator _random = RandomNumberGenerator.Create();
+        private readonly byte[] _maskingKey;
```

Comment:
Yeah, that was one of the things I was going to change, just simpler to bash out the prototype with an array an a `% 4` :).